src/v/kafka/requests/fetch_request.cc

```diff
@@ -523,37 +530,60 @@ static ss::future<> fetch_topic_partition(
  * response must be reassembled such that the responses appear in these
  * order as the partitions in the request.
  */
-static ss::future<> fetch_topic_partitions(op_context& octx) {
-    return ss::do_with(
-      octx.response_begin(), [&octx](op_context::response_iterator& resp_it) {
-          return ss::do_for_each(
-                   octx.request.cbegin(),
-                   octx.request.cend(),
-                   [&octx, &resp_it](
-                     const fetch_request::const_iterator::value_type& p) {
-                       return fetch_topic_partition(octx, p, resp_it)
-                         .then([&resp_it] { resp_it++; });
-                   })
-            .then([&octx] { octx.reset_context(); });
+static ss::future<> fetch_topic_partitions(ss::lw_shared_ptr<op_context> octx) {
+    auto resp_it = octx->response_begin();
+    std::vector<ss::future<>> fetches;
+    std::transform(
+      octx->request.cbegin(),
+      octx->request.cend(),
+      std::back_inserter(fetches),
+      [&resp_it, octx](const fetch_request::const_iterator::value_type& p) {
+          octx->fetches_in_progress++;
+          // we use gate as we may not wait for all the fetch results
+          return ss::with_gate(octx->rctx.gate(), [octx, &p, &resp_it] {
+              return fetch_topic_partition(octx, p, resp_it++).then([octx] {
+                  octx->fetches_in_progress--;
+                  octx->fetch_available.broadcast();
+              });
+          });
       });
+
+    auto f = ss::now();
+    // only wait when we still have fetches in progress and we are allowed to
+    // wait
+    if (octx->fetches_in_progress > 0 && octx->request.debounce_delay()) {
+        f = octx->fetch_available
+              .wait(
+                octx->request.debounce_delay().value(),
+                [octx] {
+                    return octx->fetches_in_progress == 0
+                           || octx->should_stop_fetch();
+                })
+              .handle_exception_type(
+                [](const seastar::condition_variable_timed_out&) {
+                    // ingore, we will refetch if we are allowed to do so.
+                });
+    } else {
+        f = ss::when_all_succeed(fetches.begin(), fetches.end());
```

Comment:
is the intention here for when_all_succeed to run the fetches in parallel? i saw a comment that they were handled in parallel in seastar, but looking at the actual code it looks like they are completed serially. i think its worth a second look if that is the intention here.

Reply:
exactly, it run the fetches in parallel, they are scheduled in the moment when they are created, there is no sequential connection between them. i.e. they do not form the continuation chain like f.e. in `ss::do_for_each`

Reply:
oh... they are run in parallel by virtue of being created via std::transform. to be clear, i think that ss::when_all_succeed waits sequentially for each to complete before moving on to the next, in case that affects anything.

Reply:
I've looked into the implementation of `when_all_succeed`. Waiting is in here:

```c++
while (pos != futures.end() && pos->available()) {
        ++pos;
}
```
`pos` is a `future` range iterator.

Reply:
doesn't this just find the first available future? after that, it does something like pos->then_wrapped(recurse).

Reply:
so they'll complete out of order, but the part that makes it parallel isn't when_all_succeed, its std::tranform.